File: pydroid/src/main/java/com/pyamsoft/pydroid/util/DialogUtil.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pydroid.util;

import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import com.pyamsoft.pydroid.helper.Checker;
import timber.log.Timber;

public final class DialogUtil {

  private DialogUtil() {
    throw new RuntimeException("No instances");
  }

  /**
   * Using the fragment manager to handle transactions, this guarantees that any old
   * versions of the dialog fragment are removed before a new one is added.
   */
  public static void guaranteeSingleDialogFragment(FragmentActivity fragmentActivity,
      @NonNull DialogFragment dialogFragment, @NonNull String tag) {
    if (fragmentActivity == null) {
      Timber.w("Cannot attach a fragment to a NULL activity. No-op");
      return;
    }

    dialogFragment = Checker.checkNonNull(dialogFragment);
    tag = Checker.checkNonNull(tag);

    if (tag.isEmpty()) {
      throw new IllegalArgumentException("Cannot use EMPTY tag");
    }

    final FragmentManager fragmentManager = fragmentActivity.getSupportFragmentManager();
    final FragmentTransaction ft = fragmentManager.beginTransaction();
    final Fragment prev = fragmentManager.findFragmentByTag(tag);
    if (prev != null) {
      Timber.d("Remove existing fragment with tag: %s", tag);
      ft.remove(prev);
    }

    Timber.d("Add new fragment with tag: %s", tag);
    dialogFragment.show(ft, tag);
  }

  /**
   * Guarantees that a fragment with the given tag is only added to the view once
   */
  public static void onlyLoadOnceDialogFragment(FragmentActivity fragmentActivity,
      @NonNull DialogFragment dialogFragment, @NonNull String tag) {
    if (fragmentActivity == null) {
      Timber.w("Cannot attach a fragment to a NULL activity. No-op");
      return;
    }

    dialogFragment = Checker.checkNonNull(dialogFragment);
    tag = Checker.checkNonNull(tag);

    if (tag.isEmpty()) {
      throw new IllegalArgumentException("Cannot use EMPTY tag");
    }

    final FragmentManager fragmentManager = fragmentActivity.getSupportFragmentManager();
    final Fragment prev = fragmentManager.findFragmentByTag(tag);
    if (prev == null) {
      dialogFragment.show(fragmentManager, tag);
    }
  }
}

File: pydroid-ui/src/main/java/com/pyamsoft/pydroid/ui/util/DialogUtil.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.pydroid.ui.util

import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentActivity
import timber.log.Timber

object DialogUtil {

    /**
     * Using the fragment manager to handle transactions, this guarantees that any old
     * versions of the dialog fragment are removed before a new one is added.
     */
    @JvmStatic
    fun guaranteeSingleDialogFragment(
        fragmentActivity: FragmentActivity?,
        dialogFragment: DialogFragment, tag: String
    ) {
        if (fragmentActivity == null) {
            Timber.w("Cannot attach a fragment to a NULL activity. No-op")
            return
        }

        if (tag.isEmpty()) {
            throw IllegalArgumentException("Cannot use EMPTY tag")
        }

        val fragmentManager = fragmentActivity.supportFragmentManager
        val ft = fragmentManager.beginTransaction()
        val prev = fragmentManager.findFragmentByTag(tag)
        if (prev != null) {
            Timber.d("Remove existing fragment with tag: %s", tag)
            ft.remove(prev)
        }

        Timber.d("Add new fragment with tag: %s", tag)
        dialogFragment.show(ft, tag)
    }
}
